desc/magnetic_fields/_current_potential.py

```diff
@@ -696,3 +902,649 @@ def nfp_loop(j, f):
+        contours_indices.append(this_contour[0])
+    # from the indices, calculate the actual zeta and theta values of the contours
+    contours_theta_zeta = [
+        np.array(
+            [
+                np.interp(contour[:, 0], np.arange(zeta_full.size), zeta_full),
+                np.interp(contour[:, 1], np.arange(theta_full.size), theta_full),
+            ]
+        ).T
+        for contour in contours_indices
+    ]
+    # reverse so that our coils start at zeta=0, this is just a choice
+    contours_theta_zeta.reverse()
+
+    numCoils = 0
+    # to be used to check closure conditions on the coils
+    ## closure condition in zeta for modular is returns to same zeta,
+    ## while for helical is that the contour dzeta = 2pi/NFP
+    zeta_diff = 2 * jnp.pi / nfp if not jnp.isclose(helicity, 0) else 0.0
+    ## closure condition in theta for modular is that dtheta = 2pi,
+    ## while for helical the dtheta = 2pi*abs(helicity)
+    theta_diff = (
+        2 * jnp.pi * jnp.abs(helicity) if not jnp.isclose(helicity, 0) else 2 * jnp.pi
+    )
+
+    if show_plots:
+        plt.contourf(
+            zeta_full_2D.T, theta_full_2D.T, jnp.transpose(phi_total_full), contours
+        )
+        plt.xlabel(r"$\zeta$")
+        plt.ylabel(r"$\theta$")
+    for j in range(N_trial_contours):
+        contour_zeta.append(contours_theta_zeta[j][:, 0])
+        contour_theta.append(contours_theta_zeta[j][:, 1])
+        # check if closed and if not throw warning
+
+        if not jnp.isclose(
+            jnp.abs(contour_zeta[-1][-1] - contour_zeta[-1][0]), zeta_diff
+        ) or not jnp.isclose(
+            jnp.abs(contour_theta[-1][-1] - contour_theta[-1][0]), theta_diff
+        ):
+            warnings.warn(
+                f"Detected a coil contour (coil index {j}) that may not be "
+                "closed, this may lead to incorrect coils, "
+                "check that the surface current potential contours do not contain "
+                "any local maxima or window-pane-like structures,"
+                " and that the current potential contours do not go across "
+                "The edges of the zeta extent used for the plotting:"
+                "the zeta=0 or zeta=2pi/NFP planes for helical coils or the"
+                " zeta=-pi/NFP and zeta=2pi+pi/NFP planes, for modular coils. "
+                "Use `show_plots=True` to visualize the contours.",
+                UserWarning,
+            )
+        numCoils += 1
+        if show_plots:
+            plt.plot(contour_zeta[-1], contour_theta[-1], "-r", linewidth=1)
+            if j > 0:
+                plt.plot(contour_zeta[-1][-1], contour_theta[-1][-1], "sk")
+            else:
+                plt.plot(
+                    contour_zeta[-1][-1],
+                    contour_theta[-1][-1],
+                    "sk",
+                    label="start of contour",
+                )
+    if not jnp.isclose(helicity, 0):
+        # right now these are only over 1 FP
```

Comment:
I dont think we can easily use NFP and virtual coils for helical coils, since we assume that every individual coil object is a closed curve. The helical coils wont close after one NFP on itself, it instead closes on itself after the entire toroidal transit.

It would require some more work to get it to work that way. Alternatively once the #844 is in, we could do this with those curves possibly.

Reply:
Could we get it to work for modular coils? (we should know when discretizing the coilset which type it is). We could leave pretty much all the logic the same, just at the very end when returning the coilset we do `CoilSet(*coils[:num_coils], NFP=NFP, ...)`

Reply:
Right now, if the coilset is modular, it will make coils over one field period, and have virtual coils with appropriate field period symmetry and stellarator symmetry.

Sorry, I had thought you were asking about helical coils, which due to the way we define curves currently, have to be defined over the entire toroidal extent, not just one field period.

Reply:
(i.e. this if block is not used at all if the coils are modular)